### PR TITLE
 Jvenstad/generalised roles 

### PR DESCRIPTION
**Pull request by jonmv**
@mpolden or @tokle please review.

### CODE REVIEW

**Comment by jonmv** [comment]
Let's not merge this before we see how the other stuff goes :)

**Comment by jonmv** [comment]
Last commit fixes a bug I introduced in the previous PR. 
The system is always part of the context, and is constant within the system. Verifying that only one system is used for all contexts in a `RoleMembership` helps enforce this.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/PathGroup.java [hunk 1]
```diff
@@ -26,33 +28,44 @@
...
-           "/application/v4/tenant/",
-           "/application/v4/tenant-pipeline/",
+    /** Paths used by tenant administrators */
+    tenantInfo("/application/v4/",
+               "/application/v4/property/",
+               "/application/v4/tenant/",
+               "/application/v4/tenant-pipeline/"),
+
+    /** Paths used by tenant administrators */
+    tenant(Matcher.tenant,
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/PathGroup.java]: Good, I was thinking of something like this, e.g. that a path group had to declare if a limited context was required or not. But this is better 👍

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/PathGroup.java [hunk 2]
```diff
@@ -4,8 +4,10 @@
 import com.yahoo.restapi.Path;
 
 import java.util.EnumSet;
+import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import java.util.regex.Pattern;
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/PathGroup.java]: Looks to be unused (hopefully)?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java [hunk 3]
```diff
@@ -52,4 +58,59 @@ public String toString() {
         RoleMembership membership(Principal user);
     }
 
+    public interface Builder {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java]: Why is this needed?

Comment (mpolden) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java]: Nm. I see it now.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java [hunk 4]
```diff
@@ -52,4 +58,59 @@ public String toString() {
...
+
+        BuilderWithRole add(Role role);
+
+        RoleMembership build();
+
+    }
+
+    public static class BuilderWithRole implements Builder {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java]: How about `RoleBuilder`?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java [hunk 5]
```diff
@@ -52,4 +58,59 @@ public String toString() {
...
+        public Builder limitedTo(TenantName tenant) {
+            roles.putIfAbsent(current, new HashSet<>());
+            roles.get(current).add(Context.limitedTo(tenant, system));
+            current = null;
+            return this;
+        }
+
+        public Builder limitedTo(TenantName tenant, ApplicationName application) {
+            roles.putIfAbsent(current, new HashSet<>());
+            roles.get(current).add(Context.limitedTo(tenant, application, system));
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java]: Consider extracting a method for these two lines that takes context as argument, to remove the repetition.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java (hunk 4 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java]: Perhaps. It's not exactly that, either, since you can still `build()` it and get a `RoleMembership`. Anyway, as you see from the usage, it shouldn't be necessary to declare variables of this kind anyway :)

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java (hunk 5 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/RoleMembership.java]: Sure.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/PathGroup.java (hunk 2 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/role/PathGroup.java]: Moved to the unit test — used to `quote(...)` the `{tenant}` et al when searching for them in the path specs.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/filter/AthenzRoleResolver.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import com.yahoo.vespa.hosted.controller.tenant.Tenant;
+import com.yahoo.vespa.hosted.controller.tenant.UserTenant;
+
+import javax.ws.rs.InternalServerErrorException;
+import java.security.Principal;
+import java.util.Optional;
+
+import static com.yahoo.vespa.hosted.controller.athenz.HostedAthenzIdentities.SCREWDRIVER_DOMAIN;
+
+public class AthenzRoleResolver implements RoleMembership.Resolver {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/filter/AthenzRoleResolver.java]: Author and description.

**Review by mpolden — APPROVED** [review]

**Comment by freva** [comment]
This broke controller
```
[2019-03-28 19:06:35.740] WARNING : configserver     Container.com.yahoo.container.di.Container 
        Failed to set up first component graph
        exception=
        java.lang.RuntimeException: When resolving dependencies of 'com.yahoo.vespa.hosted.controller.restapi.filter.ControllerAuthorizationFilter in okta-api-request-chain'
        Caused by: java.lang.IllegalStateException: No global component of class com.yahoo.vespa.hosted.controller.role.RoleMembership$Resolver to inject into component 'com.ya
hoo.vespa.hosted.controller.restapi.filter.ControllerAuthorizationFilter in okta-api-request-chain'.
        
[2019-03-28 19:06:35.741] WARNING : configserver     Container.com.yahoo.jdisc.core.ApplicationLoader   
        Exception thrown while activating application.
        exception=
        java.lang.RuntimeException: When resolving dependencies of 'com.yahoo.vespa.hosted.controller.restapi.filter.ControllerAuthorizationFilter in okta-api-request-chain'
        Caused by: java.lang.IllegalStateException: No global component of class com.yahoo.vespa.hosted.controller.role.RoleMembership$Resolver to inject into component 'com.ya
hoo.vespa.hosted.controller.restapi.filter.ControllerAuthorizationFilter in okta-api-request-chain'.
        
[2019-03-28 19:06:35.742] WARNING : configserver     stderr     java.lang.RuntimeException: When resolving dependencies of 'com.yahoo.vespa.hosted.controller.restapi.filter.Con
trollerAuthorizationFilter in okta-api-request-chain'
[2019-03-28 19:06:35.742] WARNING : configserver     stderr     Caused by: java.lang.IllegalStateException: No global component of class com.yahoo.vespa.hosted.controller.role.
RoleMembership$Resolver to inject into component 'com.yahoo.vespa.hosted.controller.restapi.filter.ControllerAuthorizationFilter in okta-api-request-chain'.
[2019-03-28 19:06:35.742] ERROR   : configserver     Container.com.yahoo.jdisc.core.StandaloneMain      
        JDisc exiting: Throwable caught: 
        exception=
        java.lang.RuntimeException: When resolving dependencies of 'com.yahoo.vespa.hosted.controller.restapi.filter.ControllerAuthorizationFilter in okta-api-request-chain'
        Caused by: java.lang.IllegalStateException: No global component of class com.yahoo.vespa.hosted.controller.role.RoleMembership$Resolver to inject into component 'com.ya
hoo.vespa.hosted.controller.restapi.filter.ControllerAuthorizationFilter in okta-api-request-chain'.
        
```

**Comment by hakonhall** [comment]
Revert fails with "Sorry, this pull request couldn’t be reverted automatically. It may have already been reverted, or the content may have changed since it was merged"
